addons/stock_account/migrations/13.0.1.1/post-migration.py

```diff
@@ -0,0 +1,221 @@
+          COALESCE(sm.price_unit, 0.0) AS price_unit,
+          sm.create_uid, sm.create_date, sm.write_uid, sm.write_date,
+          CASE WHEN sl.usage <> 'internal' AND sld.usage = 'internal' THEN 'in'
+               WHEN sl.usage = 'internal' AND sld.usage <> 'internal' THEN 'out'
+               WHEN sl.usage = 'supplier' AND sld.usage = 'customer' THEN 'dropship'
+               WHEN sl.usage = 'customer' AND sld.usage = 'supplier' THEN 'dropship_return'
+               ELSE 'other'
+          END AS move_type
+        FROM stock_move sm LEFT JOIN stock_location sl ON sl.id = sm.location_id
+            LEFT JOIN stock_location sld ON sld.id = sm.location_dest_id
+        WHERE sm.company_id = %s AND sm.product_id = %s
+            AND state = 'done'
+        ORDER BY sm.company_id, sm.product_id, sm.date, sm.id
+    """, (company_id, product_id))
+    return env.cr.dictfetchall()
+
+
+@openupgrade.logging()
+def generate_stock_valuation_layer(env):
+    company_obj = env["res.company"]
+    product_obj = env["product.product"]
+    precision_price = env["decimal.precision"].precision_get("Product Price")
+    precision_uom = env["decimal.precision"].precision_get(
+        "Product Unit of Measure"
+    )
+    companies = company_obj.search([])
+    products = product_obj.with_context(active_test=False).search([("type", "in", ("product", "consu"))])
+    all_svl_list = []
+    for product in products:
+        for company in companies:
+            _logger.info("Product:{} {} Company:{}".format(product.id, product.name, company.name))
+            history_lines = get_product_price_history(env, company.id, product.id)
+            moves = get_stock_moves(env, company.id, product.id)
+            svl_in_vals_list = []
+            svl_out_vals_list = []
+            svl_man_vals_list = []
+            svl_in_index = 0
+            h_index = 0
+            previous_price = 0.0
+            previous_qty = 0.0
+            for move in moves:
+                is_dropship = True if move["move_type"] in ("dropship", "dropship_return") else False
+                # Add manual adjusts
+                if product.cost_method in ("average", "fifo") and not float_is_zero(
+                        previous_qty, precision_digits=precision_uom):
+                    while h_index < len(history_lines) and history_lines[h_index]["datetime"] < move["date"]:
+                        price_history_rec = history_lines[h_index]
+                        if float_compare(price_history_rec["cost"], previous_price, precision_digits=precision_price):
+                            svl_vals = _prepare_man_svl_vals(
+                                price_history_rec, previous_price, previous_qty, company, product)
+                            svl_man_vals_list.append(svl_vals)
+                            previous_price = price_history_rec["cost"]
+                        h_index += 1
```

Comment:
I don`t know why do you not do it for stadart price.
If we create product and change price on product form when don`t have qty on stock, we not need create svl but have to set previous price
```suggestion
                # Add manual adjusts
                have_qty = not float_is_zero(previous_qty, precision_digits=precision_uom)
                while h_index < len(history_lines) and history_lines[h_index]["datetime"] < move["date"]:
                    price_history_rec = history_lines[h_index]
                    if float_compare(price_history_rec["cost"], previous_price, precision_digits=precision_price):
                         if have_qty:
                            svl_vals = _prepare_man_svl_vals(
                                price_history_rec, previous_price, previous_qty, company, product)
                            svl_man_vals_list.append(svl_vals)
                        previous_price = price_history_rec["cost"]
                    h_index += 1
```
Do the same for about comment

Reply:
I agree with your approach, I had added that condition because Odoo uses it (somewhere :sweat_smile:), but without analyzing it in depth because what I urgently needed to solve was the average cost method.